2-collectors/scala-stream-collector/core/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/CollectorService.scala

```diff
@@ -343,6 +347,47 @@ class CollectorService(
     if (pixelExpected) List(`Cache-Control`(`no-cache`, `no-store`, `must-revalidate`))
     else Nil
 
+  /**
+   * Determines the cookie domain to be used by inspecting the Origin header of the request
+   * and trying to find a match in the list of domains specified in the config file.
+   * @param headers The headers from the http request.
+   * @param domains The list of cookie domains from the configuration.
+   * @param fallbackDomain The fallback domain from the configuration.
+   * @return The domain to be sent back in the response, unless no cookie domains are configured.
+   * The Origin header may include multiple domains. The first matching domain is returned.
+   * If no match is found, the fallback domain is used if configured. Otherwise, the cookie domain is not set.
+   */
+  def cookieDomain(headers: Seq[HttpHeader], domains: Option[List[String]], fallbackDomain: Option[String]): Option[String] =
+    (for {
+      domainList <- domains
+      origins <- headers.collectFirst { case header: `Origin` => header.origins }
+      originHosts = extractHosts(origins)
+      matchingDomains = for {
+        host <- originHosts
+        matchingDomains <- domainList.collectFirst { case domain if validMatch(host, domain) => domain }
+      } yield matchingDomains
+      domainToUse <- matchingDomains.headOption
+    } yield domainToUse) match {
```

Comment:
you can do an `orElse(fallbackDomain)` here instead of the pattern matching but otherwise it looks great! :+1: